### DISCUSSION TITLE
tar.gz support

### DISCUSSION THREAD

**Discussion by webdawg** [solo-io/packer-plugin-arm-image#127] [Ideas]
Does this support tar.gz based images.

Like:  http://os.archlinuxarm.org/os/ArchLinuxARM-rpi-aarch64-latest.tar.gz

**Comment by webdawg** [comment]
I wrote a script to do this, but not in go.  Someone please answer this, and I can work on implementing in go, in this software.

`#!/bin/bash
#create loop device, install arch, img back up
#
###ref
#https://kiljan.org/2021/02/20/64-bit-arch-linux-arm-on-a-raspberry-pi-4-model-b-with-8-gb-of-ram/
#https://serverfault.com/questions/731632/does-nfs-and-smb-support-sparse-files
#https://stackoverflow.com/questions/1078524/how-to-specify-the-download-location-with-wget
#https://superuser.com/questions/1657887/how-to-understand-what-this-sfdisk-command-is-doing
#https://manpages.ubuntu.com/manpages/trusty/man8/sfdisk.8.html
#https://stackoverflow.com/questions/2500436/how-does-cat-eof-work-in-bash
#https://en.wikipedia.org/wiki/Here_document#Unix_shells
#https://en.wikipedia.org/wiki/Partition_type
#https://archlinuxarm.org/platforms/armv8/broadcom/raspberry-pi-4
#
#
#TODO
#F2FS, 
#https://archlinuxarm.org/forum/viewtopic.php?f=3&t=15031


#i dont like errors
set -e

###vars:
export WORKDIR=./`date +%Y.%m.%d.%H.%M.%S`-IMAGE_BUILD
export IMAGENAME=`date +%Y.%m.%d.%H.%M.%S`-IMAGE.img
export SOURCEURL="http://os.archlinuxarm.org/os/ArchLinuxARM-rpi-aarch64-latest.tar.gz"
export SOURCEFILENAME=`basename $SOURCEURL`

export PARTITION_BOOT=p1
export PARTITION_ROOT=p2

echo "Create Work Dir"
echo $WORKDIR

###do stuff:

#create work dir
echo "Making work directory."
mkdir -p $WORKDIR


#download image
echo "Getting an image from an internet server."
mkdir -p $WORKDIR/INCOMING
wget -P $WORKDIR/INCOMING $SOURCEURL 
ls $WORKDIR/INCOMING

#no testing
#exit
#end testing

echo "Create Sparse Image File"
echo $WORKDIR/$IMAGENAME
#create sparse image for loop device
dd if=/dev/zero of=$WORKDIR/$IMAGENAME bs=1 count=0 seek=8192M

#setup loop device
echo "Creating Loop Device"
export LOOPDEVICE=`sudo losetup --find --show $WORKDIR/$IMAGENAME`

echo "Loop Device Made"
echo $LOOPDEVICE

echo "Partitioning loop device."

#partition loop device
#fat32 boot
#83 - LINUX NATIVE
sudo sfdisk --wipe always $LOOPDEVICE << EOF
,256M,0c,
,,,
EOF

sudo partprobe /dev/loop0

sudo sfdisk -l $LOOPDEVICE

#format loop device
echo "Formatting loop file systems."

sudo mkfs.vfat -F 32 $LOOPDEVICE$PARTITION_BOOT
sudo mkfs.ext4 $LOOPDEVICE$PARTITION_ROOT

echo "Creating mount dir's, and mounting file systems."

mkdir -p $WORKDIR/FILESYSTEM
sudo mount $LOOPDEVICE$PARTITION_ROOT $WORKDIR/FILESYSTEM
sudo mkdir -p $WORKDIR/FILESYSTEM/boot
sudo mount $LOOPDEVICE$PARTITION_BOOT $WORKDIR/FILESYSTEM/boot

sudo bsdtar -vxpf $WORKDIR/INCOMING/$SOURCEFILENAME -C $WORKDIR/FILESYSTEM
sync

#there must be other mmcblk devices before the main sd card reader of the device
sed -i 's/mmcblk0/mmcblk1/' $WORKDIR/FILESYSTEM/etc/fstab

#edit to customize image if needed
exit


echo "Unmount, and remove loop device"
sudo umount $WORKDIR/FILESYSTEM/boot
sudo umount $WORKDIR/FILESYSTEM
sudo losetup -d $LOOPDEVICE



#do not delete
exit
#clean up for testing
echo "Clean Up For Testing"
sudo losetup -d $LOOPDEVICE
sleep 2
rm -rf $WORKDIR`

**Comment by webdawg** [comment]
I continue to work on this.  From core README of project:

![image](https://user-images.githubusercontent.com/1071152/155843262-0d25b219-2bad-43d0-aebc-1447f07db7aa.png)

![image](https://user-images.githubusercontent.com/1071152/155843401-65bb336f-9cf6-43d7-bb63-69e1925c44ed.png)

![image](https://user-images.githubusercontent.com/1071152/155843449-ae9443d0-23ea-4ca0-9148-acac8c00d2bb.png)

![image](https://user-images.githubusercontent.com/1071152/155843459-7d9c8806-6f48-4e48-9a77-ba1a8d31f56e.png)

![image](https://user-images.githubusercontent.com/1071152/155843469-ba8fd0a9-03a8-4c06-9502-260c9a92fd40.png)

![image](https://user-images.githubusercontent.com/1071152/155843945-96aefeb1-aa03-4484-8e9e-4d47f44cc8d2.png)

https://github.com/solo-io/packer-plugin-arm-image/blob/85b6001513a6727d95536eb1d84f22c49d8d5cf9/docs/builders/arm-image.mdx

![image](https://user-images.githubusercontent.com/1071152/155844016-947d2cee-f0a2-459d-8249-73e417b18e10.png)

For list of valid values, see: pkg/image/utils/images.go

![image](https://user-images.githubusercontent.com/1071152/155844353-d48b2a71-b608-4b39-a33e-ec2e9f96fad4.png)
![image](https://user-images.githubusercontent.com/1071152/155844358-9d80ae51-3bf9-4f32-890d-5a99ab0fb10a.png)
![image](https://user-images.githubusercontent.com/1071152/155844410-63ecfb27-1d1a-4a58-a986-a087bb9abdab.png)
![image](https://user-images.githubusercontent.com/1071152/155844415-d3acf215-1c18-4ea0-a861-96f0842da1a8.png)

So most of what image_type does so far is just determine what image_mounts (and there is a bit of auto detection built into it to set image_type from matching keywords in url) will be, and image_mounts is:

![image](https://user-images.githubusercontent.com/1071152/155844476-3280b177-e716-4af6-8608-adcf6c9548d8.png)

So back to this, if I set my image_type to raspberrypi I would get (which matches my partition structure):

![image](https://user-images.githubusercontent.com/1071152/155844670-25fcbe6c-2977-4014-a0de-a22c716411c0.png)

I could set image_mounts directly to, and I may try that, but it seems like the code stops if it can't detect image_type

![image](https://user-images.githubusercontent.com/1071152/155844686-340afed9-f3fe-4fc0-be8c-407ee6b6fe80.png)



So looking at the source, the one that matches the system structure of archlinux would be 


There is also some detection of file based on file extension:

![image](https://user-images.githubusercontent.com/1071152/155844586-b4ea2f0f-c459-4079-a39a-1122c3da583f.png)

![image](https://user-images.githubusercontent.com/1071152/155844740-79b07976-dcf6-47fd-b17f-8e5a677b62ea.png)

So i may be smart to set the suffix of the image to .img, but I bet that GetImageFileInCurrentDir would be ran after an unzip/etc.

 
The last thing I have to mess with is iso_url because my image is local, and my knowledge of go is limited, but from what I can see of iso_checksum:

![image](https://user-images.githubusercontent.com/1071152/155844877-21bb68af-80b1-4116-bf92-e6d47db51369.png)


I wonder if I can use file:./local/path/file.sum inside of iso_url

I also may set iso_checksum to none because I am going to be generating images on the fly.


If I wanted to mod this code to accept tars of file systems, I would want to find where we start working with loop devices possibly, interrupt using image type, and work on mounting blank loop devices, and filling them up like I did in my script.

**Comment by webdawg** [comment]
Manged to make it all the way through, and I am working on archlinux arm.  Next step for me is to figure out why networking is not working inside the image when I try to update via packer shell script:

![image](https://user-images.githubusercontent.com/1071152/155848668-aa9c7f47-bfaa-4c91-a82a-b08d7e393026.png)

Something here?:

https://github.com/solo-io/packer-plugin-arm-image/blob/master/pkg/builder/step_handle_resolv_conf.go

**Comment by webdawg** [comment]
*https://githubmemory.com/repo/solo-io/packer-builder-arm-image/issues/22

**Comment by yuval-k** [comment]
yea this does look like that resolve.conf issue.
did you try setting the `resolv-conf` in the settings? perhaps to `copy-host`?  do you have your config file available?
